src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -1671,18 +1694,37 @@ internal void ClearTransaction(Exception? disposeReason = null)
 
     #region Cancel
 
+    internal void ResetCancellation()
+    {
+        // If a cancellation is in progress, wait for it to "complete" before proceeding (#615)
+        lock (CancelLock) { }
+        if (PendingPrependedResponses > 0)
+            readingPrependedMessagesMRE.Reset();
+        Debug.Assert(readingPrependedMessagesMRE.IsSet || PendingPrependedResponses > 0);
```

Comment:
Very paranoid assertion :)

Reply:
Yep, but it's mostly to be sure that we after we've reset MRE for previous query we've actually set MRE (as otherwise we're going to be stuck on cancellation).